File: baekjoon/7662/solution.js
class PriorityQueue {
  constructor(isAsc) {
    this.queue = [];
    this.comparator = isAsc
      ? (a = Infinity, b = Infinity) => a > b
      : (a = -Infinity, b = -Infinity) => a < b;
  }

  pop() {
    const result = this.queue[0];

    this.queue[0] = this.queue[this.queue.length - 1];
    this.queue.length -= 1;

    let idx = 0;
    while (idx < this.queue.length) {
      const rIdx = (idx + 1) * 2;
      const lIdx = rIdx - 1;
      const curr = this.queue[idx];
      const lval = this.queue[lIdx];
      const rval = this.queue[rIdx];

      if (
        this.comparator(curr, lval) &&
        (this.comparator(rval, lval) || rval === lval)
      ) {
        [this.queue[idx], this.queue[lIdx]] = [
          this.queue[lIdx],
          this.queue[idx],
        ];
        idx = lIdx;
      } else if (this.comparator(curr, rval)) {
        [this.queue[idx], this.queue[rIdx]] = [
          this.queue[rIdx],
          this.queue[idx],
        ];
        idx = rIdx;
      } else {
        break;
      }
    }

    return result;
  }

  push(value) {
    let idx = this.queue.length;
    this.queue[this.queue.length] = value;

    while (idx > 0) {
      const pIdx = Math.ceil(idx / 2) - 1;
      if (this.comparator(this.queue[pIdx], this.queue[idx])) {
        [this.queue[pIdx], this.queue[idx]] = [
          this.queue[idx],
          this.queue[pIdx],
        ];
        idx = pIdx;
      } else {
        break;
      }
    }
  }

  isEmpty() {
    return this.queue.length === 0;
  }
}

class DoublePriorityQueue {
  constructor() {
    this.ascQueue = new PriorityQueue(true);
    this.descQueue = new PriorityQueue(false);
    this.checked = new Map();
  }

  push(value) {
    this.ascQueue.push(value);
    this.descQueue.push(value);

    this.checked.set(value, (this.checked.get(value) ?? 0) + 1);
  }

  pop(isAsc) {
    const queue = isAsc ? this.ascQueue : this.descQueue;

    while (!queue.isEmpty()) {
      const popped = queue.pop();
      const checkedCount = this.checked.get(popped) ?? 0;
      if (checkedCount) {
        if (checkedCount === 1) {
          this.checked.delete(popped);
        } else {
          this.checked.set(popped, checkedCount - 1);
        }
        return popped;
      }
    }
  }
}

//////
////
// input

const fs = require("fs");
const isLocal = process.platform !== "linux";
const filePath = isLocal ? "./input.txt" : "/dev/stdin";
const cases = fs.readFileSync(filePath).toString().trim().split("\n");

let idx = 0;
while (idx < cases.length) {
  let result = "";
  const n = +cases[idx++];

  for (let i = 0; i < n; i++) {
    const k = +cases[idx++];

    const queue = new DoublePriorityQueue();

    for (let j = 0; j < k; j++) {
      const [comm, value] = cases[idx++].split(" ");
      switch (comm) {
        case "I":
          queue.push(+value);
          break;
        case "D":
          queue.pop(+value === -1);
          break;
      }
    }

    // if (queue.isEmpty()) {
    //   result += "EMPTY\n";
    // } else {
    //   const min = queue.pop(true);
    //   const max = queue.pop(false) ?? min;
    //   result += `${max} ${min}\n`;
    // }
    const min = queue.pop(true);
    const max = queue.pop(false) ?? min;

    if (min !== undefined) {
      result += `${max} ${min}\n`;
    } else {
      result += "EMPTY\n";
    }
  }
  console.log(result);
}

File: baekjoon/7662/solution.py
import sys
import heapq
  
class DoublePriorityQueue:
  def __init__(self):
    self.ascQueue = []
    self.descQueue = []
    self.checked = {}

  def push(self, value):
    # print('push')
    heapq.heappush(self.ascQueue, (value, value))
    heapq.heappush(self.descQueue, (-value, value))
    # print(self.ascQueue)
    # print(self.descQueue)
    try:
      self.checked[value] = self.checked[value] + 1
    except:
      self.checked[value] = 1
    # print(self.checked)

  def pop(self, isAsc):
    # print('pop')
    queue = self.ascQueue if isAsc else self.descQueue

    while len(queue) > 0:
      popped = heapq.heappop(queue)[1]
      # print(popped)
      try:
        checkedCount = self.checked[popped]
      except:
        checkedCount = 0
      # print(checkedCount)

      if checkedCount > 0:
        if checkedCount == 1:
          del self.checked[popped]
        else:
          self.checked[popped] = self.checked[popped] - 1
        # print(self.ascQueue)
        # print(self.descQueue)
        return popped

    # print(self.ascQueue)
    # print(self.descQueue)
    return None

if sys.platform != 'linux' and sys.platform != 'linux2':
  sys.stdin = open("input.txt", "r")

n = int(sys.stdin.readline().rstrip())

for _ in range(n):
  k = int(sys.stdin.readline().rstrip())

  queue = DoublePriorityQueue()

  for _ in range(k):
    [comm, value] = list(map(
      lambda item: item[1] if item[0] == 0 else int(item[1]),
      enumerate(sys.stdin.readline().rstrip().split(' '))
    ))

    # print([comm, value])
    
    if comm == 'I':
      queue.push(value)
    elif comm == 'D':
      queue.pop(value == -1)

  min = queue.pop(True)
  max = queue.pop(False)
  if max == None:
    max = min 

  # print()
  if min != None:
    print("{max} {min}".format(max = max, min = min))
  else:
    print('EMPTY')
